File: 백준/Gold III/2457. 공주님의 정원/공주님의 정원.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();

    public static void main(String[] args) throws IOException {

        int N = Integer.parseInt(br.readLine());

        int[][] lst = new int[N][4];

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            lst[i][0] = Integer.parseInt(st.nextToken());
            lst[i][1] = Integer.parseInt(st.nextToken());
            lst[i][2] = Integer.parseInt(st.nextToken());
            lst[i][3] = Integer.parseInt(st.nextToken());
        } // 입력 완료

        Arrays.sort(lst, Comparator.comparingInt((int[] o) -> o[0]).thenComparingInt(o -> o[1]));

        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt((int[] o) -> o[0]).thenComparingInt(o -> o[1]).reversed());
        int lastMonth = 3;
        int lastDay = 1;
        int answer = 0;
        boolean isPossible = true;
        int i = 0;
        int startMonth = 0, startDay = 0, endMonth = 0, endDay = 0;
        while (i <= N) {
            if (lastMonth == 12) break;
            if (i < N) {
                startMonth = lst[i][0]; startDay = lst[i][1]; endMonth = lst[i][2]; endDay = lst[i][3];
            }
            if (i == N || startMonth > lastMonth || (startMonth == lastMonth && startDay > lastDay)) {
                if (pq.isEmpty()) {
                    isPossible = false;
                    break;
                } else {
                    answer += 1;
                    int[] temp = pq.poll();
                    lastMonth = temp[0];
                    lastDay = temp[1];
                }
            }
            if (startMonth < lastMonth || (startMonth == lastMonth && startDay <= lastDay)) {
                pq.offer(new int[] {endMonth, endDay});
            }
            i += 1;
        } // end of while

        if (lastMonth != 12) isPossible = false;

        if (isPossible)
            System.out.println(answer);
        else
            System.out.println(0);

    } // end of main

} // end of class

File: 백준/Gold III/2457. 공주님의 정원/공주님의 정원.py
"""
완탐?
n이 100_000 2^100_000

그리디?
현재 꽃이 지는 날짜보다 빨리 피는 꽃들 중 제일 늦게까지 피는 꽃
"""
import sys
import heapq

input = lambda: sys.stdin.readline().strip()

N = int(input())
lst = []
for _ in range(N):
    a, b, c, d = map(int, input().split())
    lst.append([a, b, c, d])

lst = sorted(lst)

last_month = 3
last_day = 1
answer = 0
pq = []
possible = True
i = 0
while i <= N:
    if last_month == 12:
        break
    if i < N:
        start_month, start_day, end_month, end_day = lst[i]
    if i == N or start_month > last_month or (start_month == last_month and start_day > last_day):
        if len(pq) == 0:
            possible = False
            break
        else:
            answer += 1
            last_month, last_day = heapq.heappop(pq)
            last_month = -last_month
            last_day = -last_day
            pq = []
    if start_month < last_month or (start_month == last_month and start_day <= last_day):
        heapq.heappush(pq, [-end_month, -end_day])
    i += 1

if last_month != 12:
    possible = False

if possible:
    print(answer)
else:
    print(0)